### PR TITLE
Reset Password

### LINKED ISSUE #54: Reset Password
There should be two api endpoints through which a user can reset a forgotten password.

- send an email with a link to reset the password using a token
- change the password given a valid token

### PR DESCRIPTION
**Pull request by michael-kamel**
Closes #54 

Added captcha.

### CODE REVIEW

File: src/main/java/science/icebreaker/dao/entity/AccountProfile.java [hunk 1]
```diff
@@ -98,6 +98,9 @@ public void setResearchArea(String researchArea) {
         this.researchArea = researchArea;
     }
 
+    public String getFullName() {
+        return String.format("%s %s %s", this.title, this.forename, this.surname);
```
Comment (dajenet) [inline src/main/java/science/icebreaker/dao/entity/AccountProfile.java]: There should be no space between title and forename, if the title is empty.

File: src/main/java/science/icebreaker/service/MailService.java [hunk 2]
```diff
@@ -34,17 +35,40 @@
...
         this.mailProperties = mailProperties;
+        this.hostUrl = hostUrl;
+    }
+
+    /**
+     * Sends a reset password token to the given account
+     * @param token the token to send
+     * @param account the account to send to
+     */
+    public void sendResetPasswordRequest(String token, Account account) {
```
Comment (dajenet) [inline src/main/java/science/icebreaker/service/MailService.java:61]: Maybe use the sendTemplateMail method instead.

File: src/main/java/science/icebreaker/dao/entity/ResetPasswordToken.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.OneToOne;
+
+@Entity
+public class ResetPasswordToken {
+    @Id
+    private String token;
```
Comment (dajenet) [inline src/main/java/science/icebreaker/dao/entity/ResetPasswordToken.java:13]: Using the account as id would enforce only one valid token at the same time.

File: src/main/java/science/icebreaker/dao/entity/AccountProfile.java (hunk 1 above)
Comment (michael-kamel) [reply src/main/java/science/icebreaker/dao/entity/AccountProfile.java]: Good point.

File: src/main/java/science/icebreaker/dao/entity/ResetPasswordToken.java (hunk 3 above)
Comment (michael-kamel) [reply src/main/java/science/icebreaker/dao/entity/ResetPasswordToken.java:13]: I could go with a unique index as well

**Review by vordemann — APPROVED** [review]
Nice work with the token!

**Pull request merged**